### PR TITLE
[no-op] Split Keccak constraints into smaller functions

### PR DESCRIPTION
**Pull request by querolita**
This PR simply moves the constraints of the Keccak circuit into smaller functions, each of them targeting "one type" of constraints. This is just a cosmetic change, meant to provide more fine grain access to each part of the checks for future negative tests that can attack specific constraints separately. This will come in a future PR building on top of this one.

### CODE REVIEW

File: optimism/src/keccak/interpreter.rs [hunk 1]
```diff
@@ -124,219 +124,302 @@ pub trait KeccakInterpreter<F: One + Debug + Zero> {
...
                                 None,
                                 None,
+                                Some(x),
+                                Some(q),
                             )),
                 );
-            }
-            let sponge_shifts = self.vec_sponge_shifts();
-            for i in 0..QUARTERS * WORDS_IN_HASH {
-                // In squeeze, Check shifts correspond to the 256-bit prefix digest of the old state (current)
```
Comment (volhovm) [inline optimism/src/keccak/interpreter.rs:189]: It seems that the original code has only this `QUARTERS * WORDS_IN_HASH` squeeze loop, but the new code has additional one:
```
           // In absorb, Check shifts correspond to the decomposition of the new state
            self.constrain(
                self.is_absorb()
                    * (self.new_state(i).clone()
                        - Self::from_shifts(&self.vec_sponge_shifts(), Some(i), None, None, None)),
            );
```            
over `QUARTERS * DIM * DIM`... no?

**Review by volhovm — APPROVED** [review]
Quite a complex refactoring to track -- but as long as it is a no-op it's all good. Would be nice to double check the ordering is the same as it was before (if it matters?).

File: optimism/src/keccak/interpreter.rs [hunk 2]
```diff
@@ -124,219 +124,302 @@ pub trait KeccakInterpreter<F: One + Debug + Zero> {
...
-                self.constrain(Self::either_zero(self.is_absorb(), self.is_squeeze()));
-                // Trivially, is_sponge and is_round are mutually exclusive
-            }
+            self.constrain_flags();
         }
 
         // SPONGE CONSTRAINTS: 32 + 3*100 + 16 + 6 = 354 CONSTRAINTS OF DEGREE 2
+        // - 354 constraints of degree 2
         {
-            for zero in self.sponge_zeros() {
```
[the remark this answers is not in the document]
Comment (querolita) [reply optimism/src/keccak/interpreter.rs:161]: the ordering of constraints does not matter 😉 

correct, inside `constrain_sponge()` you have `constrain_absorb()`

File: optimism/src/keccak/interpreter.rs (hunk 1 above)
Comment (querolita) [reply optimism/src/keccak/interpreter.rs:189]: The squeeze loop is over `QUARTERS * WORDS_IN_HASH` but the one for absorb is over `QUARTERS * DIM * DIM`

**Pull request merged**